Backend style guide: Document database patterns (#30219)
grafana/grafana: contribute/style-guides/backend.md

  allocating heap memory)
* You might *need* `nil` to tell if a variable isn't set, although usually it's better to use the type's zero
  value to tell instead

## Database

> 

Reviewer: "Avoid stacked headings, which is following one heading with another heading." See https://github.com/grafana/grafana/blob/master/contribute/style-guides/documentation-style-guide.md#heading-donts

> 

Reviewer 2: This topic has the following sections.

### Foreign keys

While they can be useful, we don't generally use foreign key constraints in Grafana, for historical and
technical reasons. See this [comment](https://github.com/grafana/grafana/issues/3269#issuecomment-383328548) by Torkel 

> for context.

Reviewer: While they can be useful, for historical and
technical reasons we do not generally use foreign-key constraints in Grafana. For context, refer to this [comment](https://github.com/grafana/grafana/issues/3269#issuecomment-383328548).

Is there a reason any specific person needs to be named explicitly (in this case Torkel)?

Author: @osg-grafana It's useful for us working on the code to know who is behind this architectural decision, and how it came about.

> While they can be useful, we don't generally use foreign key constraints in Grafana, for historical and
> technical reasons. See this [comment](https://github.com/grafana/grafana/issues/3269#issuecomment-383328548) by Torkel 
> for context.

Reviewer 2: ```suggestion
While they can be useful, we don't generally use foreign key constraints in Grafana, for historical and
technical reasons. Refer to this [comment](https://github.com/grafana/grafana/issues/3269#issuecomment-383328548) for context.
```

Author: @achatterjee-grafana All your suggestion does here is join two lines. Am I breaking a style guideline wrt. Markdown line length? Lines shorter than a certain limit are more practical to read, especially when diffing.

Reviewer 2: It is an internal document, a style guide. If the break helps you read, we can go with that.

Author: I don't really understand. I was wondering if you suggested the change, because a style guideline dictates a longer line length.

Author: I'm not sure if we have a standard for Markdown line lengths, but I think they should generally be kept shorter than a certain limit (~100-120 chars).

Reviewer 3: there is one guideline seems:
`Basic conventions for Markdown files
Wrap all lines at 80 characters. Denote bold text using the asterisk format: **bold text**`

Author: @ying-jeanne Do you have a link for that?

Reviewer 2: We do not anything specific in the [markdown style guide](https://github.com/grafana/grafana/blob/master/contribute/style-guides/documentation-markdown-guide.mdl).

Author: If it's a _standard_ Markdown convention you're referring to, we're probably not following it. 80 sounds a bit conservative anyhow.

Author: @achatterjee-grafana In that case, we can resolve this.

Reviewer 3: I doubt there is no offical guideline for it, seems a lot of projects put 80-100 characters https://www.google.com/search?q=official+markdown+line+length&rlz=1C1CHBF_enFR925FR925&sxsrf=ALeKk03doieh4V3RkA-KwYGpl138driGIg:1610464412361&ei=nLz9X_LSFcOaa7S2qFg&start=0&sa=N&ved=2ahUKEwjykOO515buAhVDzRoKHTQbCgs4ChDy0wN6BAgJEDU&biw=1920&bih=937

### Unique columns

> If a column, or column combination, should be unique, add a corresponding uniqueness constraint through a migration.

Reviewer 2: ```suggestion
If a column, or column combination, is unique, add a corresponding uniqueness constraint through a migration.
```

Author: @achatterjee-grafana "should be unique" is the correct wording here. Constraints are added to catch if the rule is broken.

Reviewer 2: Thanks for the clarification.

  allocating heap memory)
* You might *need* `nil` to tell if a variable isn't set, although usually it's better to use the type's zero
  value to tell instead

## Database

In database related code, we follow certain patterns.

### Foreign keys

> While they can be useful, we don't generally use foreign key constraints in Grafana, for historical and

Reviewer: I will add explicit rationale next time:

I rephrased this so that if you took the content between commas out, the sentence still makes sense.
Also, "See this" should really be "Refer to" for consistency across docs.
Otherwise, we have "See", "Check out", "Read", "Refer to" etc to mean the same thing."

Reviewer 4: @osg-grafana I also think @aknuds1 missed that there was a suggestion (and not just a quote) in addition to your question. If you use the suggestion code block like 👇 (or press the `[±]` button in GitHub's Markdown editor) it'll be clear that you're making a suggestion.

```markdown
    ```suggestion
    While they can be useful, we don't generally use domestic key constraints in Grafana, for historical and
    ```
```

turns into

```suggestion
While they can be useful, we don't generally use domestic key constraints in Grafana, for historical and
```

Reviewer: Oh, this is brilliant. Thank you!
